W-16325926 Adding steps for each type of connected app (#378)
mulesoft/docs-access-management: modules/ROOT/pages/connected-apps-developers.adoc

Anypoint Platform provides a https://anypoint.mulesoft.com/exchange/portals/anypoint-platform/[comprehensive set of APIs] to extend functionality.

The Connected Apps feature enables you to:

* Delegate access to applications that use APIs to interact with Anypoint Platform programmatically. 
* Build CI/CD pipelines or productize additional third-party use cases on top of Anypoint Platform.
* Enable or disable clients and view client usage. For example, you can see how many organizations are using the app and how many users have access.

== Authentication

The Connected Apps feature supports OAuth 2.0 and OpenID Connect, so you must ensure your app can use tokens to request access to user data.

See the information in the Open ID configuration https://anypoint.mulesoft.com/accounts/api/v2/oauth2/.well-known/openid-configuration[discovery endpoint] to find the supported OpenID Connect capabilities. 

To make the authentication experience consistent for end users, third-party developers with their own websites or web apps can allow users to sign in using the *Sign in with Anypoint Platform* option.

[[application-allowlist-dev]]
== Application Allowlist

> Organization administrators can enable an allowlist to control which apps can access the organization's users' data.

Reviewer: ```suggestion
Organization administrators can enable an allowlist to control which apps can access the organization's user data.
```

Reviewer: Probably not correct, but the double possessives made my eye twitch.

> When your app is added to an allowlist, only the set of scopes associated with the app are reviewed and approved by the administrator. If you add new scopes to your app at a later date, you are unable to request those scopes from users until the admin reapproves your app.

Reviewer: ```suggestion
When your app is added to an allowlist, only the set of scopes associated with the app are reviewed and approved by the administrator. If you add new scopes to your app later, you’re unable to request those scopes from users until the admin reapproves your app.
```

When you create connected apps, keep in mind: 

* Your organization can own up to 2000 connected apps at the root organization or business group levels.
* Each connected app can have up to 1000 scopes.
* Your organization owns your app. Any organization administrator of the root organization or business group in which it resides can control it.
* A unique `clientID` is automatically assigned and can’t be changed. The `clientSecret` has no expiration date by default but can be changed.

== Before You Begin

* You need the Organization Administrator permission at the level you want to create connected apps and specify access scopes. 
* Determine whether the app is internal to your company or is a third-party app that can be distributed outside of your organization.

[[create-connected-app]]
== Create a Connected App That Acts on Behalf of a User

include::partial$include-nav-connected-apps.adoc[]

. In *Owned Apps*, click *Create App*.
. In the *Create App* page, enter a unique name for the application.

> . For *Type*, select *App acts on behalf of a user*. 

Reviewer 2: First, I *love* that you're splitting out separate sections for "on Its Own Behalf" vs "on Behalf of a User". It is so much more readable!

I do think, though, that the first few steps should be consistent between the two sections, since the flow doesn't diverge until you select the grant type. In particular, it feels strange that there are different starting points between the different sets of steps; the user app steps here assume you're already in Connected Apps, whereas the client app steps below don't even assume that you're in Access Management yet.

If nothing else, I think it makes a lot more sense to steer the user towards "Connected Apps" as below, instead of steering them to "Owned Apps". If I was simply told "In Owned Apps, click Create App," I think it would take me awhile to figure out that Owned Apps was a tab label _within_ the Connected Apps page, and not a top-level navigation option. Up to you whether or not it's worth mentioning Owned Apps at all, I just think it should be contexualized if mentioned.

Also, it's probably worth at least mentioning that a Business Group can be selected before creating an app, although I presume it will be left untouched in the majority of cases, to simply default to the root org (or whatever org a non-root-org admin can access).

=== Refresh Token

Use the `refresh_token` grant type to grant access to user data when the user isn’t signed in. Refresh tokens expire after 90 days. Access tokens expire after the user’s session expires.

To work with OpenID Connect, apps have to exchange refresh tokens for new access tokens after the initial authorization to continue accessing user data.

If the app is using the authorization code grant type, the `offline_access` scope is required to use refresh tokens after the original access tokens expire.

To request a new access token, send a request to the token endpoint using the `refresh_token` grant type. When authentication is successful, a refresh token is exchanged for an access token.

=== Password

Use the `password` grant type when the application exchanges the user’s username and password for an access token.

> Because the app has to collect the user’s credentials and the `password` grant type's limitation in using mechanisms for things like multi-factor authentication, using this grant type isn't recommended.

Reviewer 2: Good idea trying to smooth out this wording, but I still find it a little hard to parse, so I think you can take it even further. I'll throw a suggestion in the mix, but you can take it or leave it.

```suggestion
This grant type is not recommended, as it exposes user credentials directly to the connected app. It also precludes the use of additional security measures such as multi-factor authentication.
```

| The authentication server. For example, the authentication server for the US region is `https://anypoint.mulesoft.com/accounts/api/v2/oauth2/token`.

|===
. The app calls the authorization server with the generated assertion.

=== Client Credentials

Use the `client_credentials` grant type for machine-to-machine authentication where a specific user’s permission to access data isn't required.

The `client_credentials` grant type uses the `clientId` and `clientSecret` of the client to authorize and access protected data.

> CI/CD pipelines using service accounts to programmatically call Anypoint Platform can be configured to use connected apps to prevent service disruptions when configuring xref:multi-factor-authentication.adoc[multi-factor authentication].

Reviewer 2: It feels like this belongs with the "Create a Connected App That Acts on Its Own Behalf". Although I can see how it technically belongs under this "Grant Types and Token Flow" section as well. Maybe just add a link between the sections?

* `openid`: Read-only access to the user's username and unique Anypoint ID.
* `profile`: Read-only access to the user's Anypoint profile.
* `email`: Read-only access to the user's email.
* `offline_access`: Access to the user's data when they're not signed in.
* `openid:google_wif`: Enables you to configure Anypoint as an IdP for Google Cloud by providing an `id_token` and an access token during the OAuth2 token flow for `client_credentials` connected apps. This ID token conforms to the IdP requirements as outlined in the https://cloud.google.com/iam/docs/workload-identity-federation-with-other-providers[GCP's WIF requirements].

[NOTE]

> Apps that use the `client_credentials` grant type always have the `profile` scope assigned to them implicitly. The data is returned by queries using the scope that corresponds to the user who created the connected app.

Reviewer 2: It's a little ambiguous as to how to resolve the grammatical error here, but I read "The data returned by queries" as the subject, and "using [vf. uses]" as the improperly-used verb.

```suggestion
Apps that use the `client_credentials` grant type always have the `profile` scope assigned to them implicitly. The data returned by queries uses the scope that corresponds to the user who created the connected app.
```
